py/malloc.c

```diff
@@ -248,12 +257,7 @@ void *m_tracked_calloc(size_t nmemb, size_t size) {
     size_t n = m_tracked_count_links(&nb);
     DEBUG_printf("m_tracked_calloc(%u, %u) -> (%u;%u) %p\n", (int)nmemb, (int)size, (int)n, (int)nb, node);
     #endif
-    if (MP_STATE_VM(m_tracked_head) != NULL) {
-        MP_STATE_VM(m_tracked_head)->prev = node;
-    }
-    node->prev = NULL;
-    node->next = MP_STATE_VM(m_tracked_head);
-    MP_STATE_VM(m_tracked_head) = node;
+    m_tracked_node_insert(&MP_STATE_VM(m_tracked_head), node);
```

Comment:
Do we need to worry about thread safety for any of these tracked node linked list operations? I think they're always called holding the GIL on GIL ports, but on non-GIL threaded ports (i.e. rp2) there's nothing stopping two threads from calling in here at the same time for unrelated objects, is there?

(Comment applies to the existing functions as much as the changes in this PR, although now we also potentially traverse a linked list during a GC pass.)

Reply:
Yes, I think you're right that it needs to be protected.  I guess everything in `mp_state_vm_t` does (`mp_state_mem_t` should already be protected by `gc_mutex`, and `mp_state_thread_t` is by definition per-thread so doesn't need protecting).

Reply:
Understood. Will look at fixing in a separate PR.

Reply:
Have submitted this draft:
https://github.com/micropython/micropython/pull/16071

Is stacked on top of this PR, currently.